### ISSUE TITLE
WithGiven operations do not consider the cache of Non-WithGiven-operations

### ISSUE THREAD

**Issue by zickgraf** [homalg-project/CAP_project#582]
Consider the following demo code:
```
LoadPackage( "CAP" );

cat := CreateCapCategory( );
myobj := SymmetricGroup( 3 );
AddObject( cat, myobj );
mymor := GroupHomomorphismByImages( myobj, myobj );
AddMorphism( cat, mymor );

AddKernelObject( cat, function( a )
    
    Display( "KernelObject" );
    
    return myobj;
    
end );

AddKernelEmbeddingWithGivenKernelObject( cat, function( a, b )
    
    Display( "KernelEmbeddingWithGivenKernelObject" );
    
    return mymor;
    
end );

AddKernelEmbedding( cat, function( a )
    
    Display( "KernelEmbedding" );
    
    return mymor;
    
end );


KernelEmbedding( mymor );
KernelEmbedding( mymor );
KernelEmbeddingWithGivenKernelObject( mymor, myobj );
```

Due to the caches I would expect that the kernel embedding of `mymor` is only computed a single time, that is, I expect the output:
```
KernelEmbedding
```

However, `KernelEmbeddingWithGivenKernelObject` recomputes the kernel embedding of `mymor`, i.e., I get the output:
```
KernelEmbedding
KernelEmbeddingWithGivenKernelObject
```

I think KernelEmbeddingWithGivenKernelObject should consider the cache of KernelEmbedding before doing an actual computation.

Note: Of course this example is artificial, but this actually occurs in real code in the following setting:

Assume that KernelLift is derived from KernelEmbedding and LiftAlongMonomorphism. Let alpha and beta be morphisms. Now:
1. Compute KernelEmbedding( alpha ). In particular, KernelObject( alpha ) is automatically set by the post function.
2. Compute KernelLift( alpha, beta ).
Since KernelObject( alpha ) is known, KernelLift( alpha, beta ) redirects to KernelLiftWithGivenKernelObject( alpha, beta, KernelObject( alpha) ). Since KernelLiftWithGivenKernelObject is derived from KernelEmbeddingWithGivenKernelObject by the derivation pair, now KernelEmbeddingWithGivenKernelObject is called and KernelEmbedding is recomputed.

I think this setting might be quite common for computations in MatrixCategory resp. CategoryOfRows.

**Comment by zickgraf** [comment]
> In particular, we'll need to take care not to run through caches twice.

Good point. I think it should be safe to create a redirect function which simply goes through the cache of `KernelEmbedding` and if there is a hit, returns the value. However, there might be pitfalls I'm not aware of.

> is it correct that a simple "solution" to the problem is to delete the primitively given `KernelEmbeddingWithGivenKernelObject` from the matrix category

Yes, that is correct. In that case everything would go through `KernelEmbedding`, and in particular its cache would be respected. And in the matrix category, `KernelEmbedding` and `KernelEmbeddingWithGivenKernelObject` only differ by a call to `VectorSpaceObject`, so it could actually make sense to only install `KernelEmbedding`.

In ModulesPresentations there would actually be a real difference between the two versions, but `KernelEmbeddingWithGivenKernelObject` is only given as a comment. It seems like the solution there is to compute the kernel object lazily (i.e., via `LazySyzygiesOfRows`).

**Comment by sebastianpos** [comment]
> And in the matrix category, `KernelEmbedding` and `KernelEmbeddingWithGivenKernelObject` only differ by a call to `VectorSpaceObject`, so it could actually make sense to only install `KernelEmbedding`.

Right, so let's drop `KernelEmbeddingWithGivenKernelObject` in that case (same is true in Rows/Cols in the abelian case) as a first step.

Would it be okay for you if we leave it like that at the moment (this issue can still be open)? Or did you find other instances where such a recomputation happend?

**Comment by zickgraf** [comment]
That would be fine for now, I have not found other instances. I will prepare a PR.

**Comment by zickgraf** [comment]
A further question regarding this topic: Do I have to fulfil the specification
```
IsEqualForMorphisms( KernelEmbedding( alpha ), KernelEmbeddingWithGivenKernelObject( alpha, KernelEmbedding( alpha ) ) )
```
if I install both `KernelEmbedding` and `KernelEmbeddingWithGivenKernelObject`? I assume the answer is "obviously yes" :D But I could not find this specification anywhere and thus wanted to make sure.

**Comment by sebastianpos** [comment]
The answer is yes :)

**Comment by zickgraf** [comment]
> The answer is yes :)

Okay, then for the record/future reference: this issue in particular implies that if I use `IsIdenticalObj` for the equality of morphisms and rely on crisp caching, currently I may not install both a WithGiven and a non-WithGiven version of an operation since the cache of the non-WithGiven version might not be considered.

**Comment by sebastianpos** [comment]
I agree. 

Do you have an example in mind where one actually wants to implement both versions while having `IsIdenticalObj` as equality for morphisms?

**Comment by zickgraf** [comment]
No, I don't, I just try to understand all of this in detail and try to write it down in case I need it in the future.
